### PR TITLE
Infer function parameter types from partially eta-expanded methods

### PR DESCRIPTION
**Pull request by adriaanm**
This improves inference of function parameter types for partially eta-expanded methods (i.e., when only some arguments are omitted using `_`)

Example from the test, given `def repeat(x: Int, y: String) = y * x`, partial eta-expansion recovers fun param types from the definition of the selected method:
```
val sayAaah = repeat(_, "a") // place holder syntax
val thrice = x => repeat(3, x) // explicit version
val repeatFlip = (x, y) => repeat(y, x) // explicit version, two params
```

Back ported to 2.12.8 as #7340
See https://github.com/lampepfl/dotty/issues/2570

### CODE REVIEW

File: src/compiler/scala/tools/nsc/typechecker/Typers.scala [hunk 1]
```diff
@@ -2991,27 +2991,62 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
...
+              if (!vd.tpt.isEmpty) Right(vd.tpt.tpe)
+              else Left(args.indexWhere {
+                case Ident(name) => name == vd.name
+                case _           => false // TODO: i think we need to deal with widening conversions too??
+              })
+            }
+
+          // If some of the vparams without type annotation was not applied to `meth`,
+          // we're not going to learn enough from typing `meth` to determine them.
+          if (formalsFromApply.exists{ case Left(-1) => true case _ => false }) EmptyTree
```
Comment (lrytz) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala]: `.exists(_ == Left(-1))` should work, or you mind the `Left` allocation?

File: src/compiler/scala/tools/nsc/typechecker/Typers.scala [hunk 2]
```diff
@@ -2991,27 +2991,62 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
...
+          // This list specifies a way to compute the expected parameter type for each of our function's arguments in order.
+          // Either we already know it, and then we have a Type, or we don't, and then it's an index `idx` into
+          // the arguments passed to `meth`, so we can derive it from its MethodType
+          // (based on where the function's parameter is applied to `meth`)
+          val formalsFromApply =
+            vparams.map { vd =>
+              if (!vd.tpt.isEmpty) Right(vd.tpt.tpe)
+              else Left(args.indexWhere {
+                case Ident(name) => name == vd.name
+                case _           => false // TODO: i think we need to deal with widening conversions too??
```
Comment (lrytz) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala]: I don't understand this comment..

**Review by lrytz — APPROVED** [review]
Nice!

File: src/compiler/scala/tools/nsc/typechecker/Typers.scala (hunk 1 above)
Comment (adriaanm) [reply src/compiler/scala/tools/nsc/typechecker/Typers.scala]: True, or even `.contains(Left(-1))`. I was probably experimenting with fancier logic, but in the end it came out simpler.

File: src/compiler/scala/tools/nsc/typechecker/Typers.scala (hunk 2 above)
Comment (adriaanm) [reply src/compiler/scala/tools/nsc/typechecker/Typers.scala]: Not sure anymore (I wrote that a long time ago), but I think there are certain cases where an `Apply` might wrap the `Ident` in some other tree after `adapt`.

Comment (lrytz) [reply src/compiler/scala/tools/nsc/typechecker/Typers.scala]: Ah OK. In any case it's safe to keep the current version - trying to idenitfy conversions might lead to false positives, and therefore bugs. So currently we (possibly) err on the safe side.

Comment (adriaanm) [reply src/compiler/scala/tools/nsc/typechecker/Typers.scala]: probably https://github.com/scala/bug/issues/9738?

Comment (adriaanm) [reply src/compiler/scala/tools/nsc/typechecker/Typers.scala]: yep, I'll clarify the comment

**Pull request merged**